Class/service

CarlosCraveiro opened the pull request:
<!--PS: 🇧🇷/🇬🇧 It's ok to write your PR in portuguese 😃-->

## What type of PR is this? (check all applicable)
- [x] ✨ Feature
- [x] 🐞 Bug Correction
- [ ] 📝 Documentation Update
- [ ] 🚩 Other

# Changes:
<!-- Describe the main changes and the **expected changes** (if aplicable) -->
Implementando duas classes : IService e Service-Proxy

# Testing
<!--(if aplicable) Tell us how to replicate the expected behaviour.-->
Classes compliladas e testadas. (Aviso! Caso desejem compilar utilizando o CMake do repositório, sugiro adicionarem como dependencias apenas  "service" e "service-proxy", para evitar erros de compilação relacionados aos demais arquivos)

CarlosCraveiro committed:
Adding the Routine interface to the routine.cpp and routine.hpp archives

CarlosCraveiro committed:
Fixing compiling problems with routine.hpp

CarlosCraveiro committed:
adding the routine-handler functionality

CarlosCraveiro committed:
solving a mutex problem in routineModule

CarlosCraveiro committed:
Adding checkStatus and update functionalities to the routine-handler

CarlosCraveiro committed:
Starting the service.hpp

CarlosCraveiro committed:
Merge branch 'develop' into class/service

CarlosCraveiro committed:
Refactoring the Repo based in new decisions

CarlosCraveiro committed:
Removing an old not needed anymore class

CarlosCraveiro committed:
Merge branch 'develop' into class/service

CarlosCraveiro committed:
Implementing the IService

CarlosCraveiro committed:
Implementing the service-proxy

CarlosCraveiro committed:
fixing inheritance problems in IService

CarlosCraveiro committed:
implementing an alpha version of class ServiceProxy

CarlosCraveiro committed:
Merge branch 'develop' into class/service

CarlosCraveiro committed:
adding JavaDoc

CarlosCraveiro committed:
Adding JavDocs for service-proxy

mairacanal commented on src/classes/service-handler/service-handler.hpp:
```diff
@@ -0,0 +1,26 @@
+#include "../routine-handler/routine-handler.hpp"
+
+class Service {
+	public:
+		 std::vector<RoutineHandler> routines;
+		 std::vector<EndpointHandler> endpoints;
```
O ideal é que atributos de classes sejam privados e também sigam a nomenclatura comum, ou seja, `m_routines`e `m_endpoints`.

mairacanal commented on src/classes/service-handler/service-handler.hpp:
```diff
@@ -0,0 +1,26 @@
+		 std::vector<RoutineHandler> routines;
+		 std::vector<EndpointHandler> endpoints;
+	public:
+		 nlohmann::json getServiceStatus(void);
+		nlohmann::json updateAll(nlohmann::json updateList);
+		nlohmann::json registerEndpoints(void);
+		nlohmann::json registerRoutines(void);
```
Colocar as trailing functions , considerando a adoção generalizada.

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,114 @@
+        UNINITIALIZED,
+        INITIALIZED,
+        RUNNING,
+        STOPED,
```
Está escrito errado: `STOPPED`

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,73 @@
+ServiceProxy::ServiceProxy(IService& service, nlohmann::json configs)
+    : m_innerService(service)
+{
+    m_dependencies.data = configs; //DETAILS ABOUT DEPENDENCIES
```
Crie um construtor para o SafeJson e, então, passe o `configs` por meio do construtor padrão do C++

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,114 @@
+    /* structure made to allow the use of the service-proxy 
+	 * conf-file in multiples threads.
+	 */
+    struct SafeJson {
+        nlohmann::json data;
+        std::mutex mtx;
+    };
```
Criar um construtor

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,114 @@
+ 	 * @brief	function that encapsulates the class @ref IService
+	 * 			to be runned in a thread as explicit in @ref run
+ 	 */
+    static void servicePod(IService& service, Status& status);
```
Eu achei que o nome `servicePod`é pouco expressivo. Talvez poderia mudar para um nome mais significativo.

Grillo-0 replied:
Uma sugestão: `serviceLifeCycle `

mairacanal commented on src/classes/service/service.hpp:
```diff
@@ -0,0 +1,57 @@
+	 * @param	serviceId std::string that is the id related
+	 * 			to the instantiated class.
+	 */
+	explicit IService(std::string serviceId)
```
Passar por referência no argumento (lembre-se que temos apenas 512mb)

Math-42 commented on src/classes/service-handler/service-handler.hpp:
```diff
@@ -0,0 +1,26 @@
+/*
+ * service.hpp
+ *
```
Remover esse header

Math-42 commented on src/classes/service/service.cpp:
```diff
@@ -0,0 +1,10 @@
+* Author: Carlos Craveiro (@CarlosCraveiro)
+* Created On: September 11, 2021
+* 
+*/
```
remover header

Math-42 commented on src/classes/service/service.hpp:
```diff
@@ -0,0 +1,57 @@
+    
+	/**
+	 * @brief	An interface function to be implemented by 
+	 * 			the user to make the "setup"/configuration 
```
deixar os comentários alinhados com os asteriscos

Math-42 commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,114 @@
+class ServiceProxy {
+private:
+    /* a reference for a Service implemented by the user */
+    IService& m_innerService;
```
No desing pattern de proxy, é comum chamar a variável que referencia o objeto real como "real". Sugiro alterar para m_realService.

Math-42 commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,73 @@
+    service.setup();
+    status.setState(Status::RUNNING);
+
+    while (status.getState() != Status::STOPED) {
```
Aqui não seria melhor se fosse ?
```suggestion
    while (status.getState() == Status::RUNNING) 
```
Existem outros estados além desses e o único que permite que ela esteja de fato rodando é o running, não?

CarlosCraveiro replied:
A adoção do `!= Status::STOPPED` tem o objetivo de reservar ao status STOPPED - e somente a ele - o direito de encerrar "rotinas". Assim, a idea era que cada "state" tivesse um objetivo no controle da thread, permitindo ao programador por meio da requisição de status, saber exatamente o que "está acontecendo"/"aconteceu" com cada um de seus "services".

CarlosCraveiro committed:
deleting unnecessary header

Grillo-0 commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,73 @@
+{
+    std::thread thread(servicePod, std::ref(m_innerService), std::ref(m_status));
+    m_status.setState(Status::INITIALIZED);
+    std::swap(thread, m_innerThread);
```
Esse é  o melhor jeito de criar a m_innerThread?
Essa criação dessa thread auxilar me parece estranho.

Grillo-0 commented on src/classes/service-handler/service-handler.hpp:
```diff
@@ -0,0 +1,26 @@
+
+		void buildEndpoint(/*IEndpoint& newEndpoint*/);
+		void buildRoutine(IRoutine& newRoutine);
+};
```
acertar identação e seguir o estilo Webkit

Grillo-0 commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,73 @@
+    m_state = newState;
+}
+
+void ServiceProxy::servicePod(IService& service, Status& status)
```
```suggestion
void ServiceProxy::servicePod()
```

Grillo-0 commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,73 @@
+
+void ServiceProxy::run()
+{
+    std::thread thread(servicePod, std::ref(m_innerService), std::ref(m_status));
```
Tente fazer isso
```suggestion
    std::thread thread(&ServiceProxy::servicePod, this);
```
https://stackoverflow.com/questions/17472827/create-thread-inside-class-with-function-from-same-class

Grillo-0 commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,114 @@
+public:
+    /*enum for the possible service-thread states*/
+    enum stateT {
+        MISSINGDEPENDENCIES = 0,
```
```suggestion
        MISSING_DEPENDENCIES = 0,
```

Grillo-0 commented on src/classes/service/service.hpp:
```diff
@@ -0,0 +1,57 @@
+
```
renomear o arquivo para iservice.hpp

Grillo-0 commented on src/classes/service/service.cpp:
```diff
@@ -0,0 +1,3 @@
+
```
precisa desse arquivo já que a classe é puramente virtual?

CarlosCraveiro committed:
fix: STOPPED syntax

CarlosCraveiro committed:
improve: SafeJson with standard C++ inline constructor

CarlosCraveiro committed:
Adding get status methods to ServiceHandler

CarlosCraveiro committed:
service-proxy: Beginning the refactor of class' header-file

CarlosCraveiro committed:
IService: updating to allow "override" detection

CarlosCraveiro committed:
ServiceProxy: refactoring ".h" with state design pattern

CarlosCraveiro committed:
ServiceProxy: finishing refactoring process

CarlosCraveiro committed:
ServiceProxy: Fixing bugs, leak errors, etc...

CarlosCraveiro committed:
ServiceHandler: Refactoring to combine with the new ServiceProxys

CarlosCraveiro committed:
IService: fixing override bug

mairacanal commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,48 @@
+ServiceHandler::ServiceHandler(nlohmann::json servicesConfigs)
+{
+    //JSON DE CONFIGS
+    for (nlohmann::json::iterator it = servicesConfigs["proxys"].begin(); it != servicesConfigs["proxys"].end(); it++) {
```
Mudar o loop para uma versão mais moderna com iteradores

mairacanal commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,48 @@
+    for (nlohmann::json::iterator it = servicesConfigs["proxys"].begin(); it != servicesConfigs["proxys"].end(); it++) {
+        std::map<std::string, Status::stateT> depsMap;
+
+        for (nlohmann::json::iterator itr = it.value().begin(); itr != it.value().end(); itr++) {
```
Mudar o loop para uma versão mais moderna com iteradores

mairacanal commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,48 @@
+void ServiceHandler::buildServiceProxy(IService& userService, ServiceProxy::proxyT proxyType)
+{
+    if (proxyType == ServiceProxy::STATIC_SERVICE) {
+        m_serviceMap.insert(std::pair<std::string, ServiceProxy*>(userService.m_serviceId, new staticService::Proxy(userService, m_proxyDepsMap.at(userService.m_serviceId))));
```
Tenta deixar uma linha um pouco menor, para melhorar a legibilidade.

mairacanal commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,48 @@
+{
+    std::map<std::string, ServiceProxy*>::iterator itr;
+    nlohmann::json allStatus;
+    for (itr = m_serviceMap.begin(); itr != m_serviceMap.end(); itr++) {
```
Mudar o loop para uma versão mais moderna com iteradores

mairacanal commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,48 @@
+
+ServiceHandler::~ServiceHandler() {
+    std::map<std::string, ServiceProxy*>::iterator itr;
+    for (itr = m_serviceMap.begin(); itr != m_serviceMap.end(); itr++) {
```
Mudar o loop para uma versão mais moderna com iteradores

mairacanal commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,48 @@
+}
+
+ServiceHandler::~ServiceHandler() {
+    std::map<std::string, ServiceProxy*>::iterator itr;
```
Apaga essa variável

mairacanal commented on src/classes/service-handler/service-handler.hpp:
```diff
@@ -0,0 +1,22 @@
+#include <string>
+
+class ServiceHandler {
+//private:
```
Apagar o comentário

mairacanal commented on src/classes/service-handler/service-handler.hpp:
```diff
@@ -0,0 +1,22 @@
+    auto getAllProxyStatus() -> nlohmann::json;
+
+    void buildServiceProxy(IService& userService, ServiceProxy::proxyT proxyType);
+    explicit ServiceHandler(nlohmann::json servicesConfigs);
```
Coloca o construtor no começo para facilitar a leitura.

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+{
+    std::map<std::string, Status::stateT>::iterator itr;
+
+    for (itr = depsMap.begin(); itr != depsMap.end(); itr++) {
```
Mudar o loop

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+
+    for (itr = depsMap.begin(); itr != depsMap.end(); itr++) {
+
+        //m_depsMap.emplace(itr->first, Dependencie(itr->second, Status::UNKNOWN));
```
Apagar o comentário

CarlosCraveiro replied:
Resolvido!!

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+
+void staticService::MissingDependencies::allFine()
+{
+
```
Apaga a linha

CarlosCraveiro replied:
Não entendi tão bem, mass apliquei o ClangFormat.

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,256 @@
+
+class Status {
+public:
+    enum stateT { MISSING_DEPENDENCIES = 0,
```
Péssimo nome de enum

CarlosCraveiro replied:
Nome de enum alterado para state_t, quis manter o "t"zinho, acho que deixa mais claro que trata-se de um "tipo artificial".

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+void staticService::Uninitialized::somethingIsMissing()
+{
+    // Unexpose/Hide the endpoint on the DBUS and THEN..
+    m_upperProxy->changeState(MISSING_DEPENDENCIES);
```
Colocar o `Status::`no MISSING_DEPENDENCIES

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+
+    std::map<std::string, ProxyConfigs::Dependencie>::iterator itr;
+
+    for (itr = m_proxyConfigs.m_depsMap.begin(); itr != m_proxyConfigs.m_depsMap.end(); itr++) {
```
Mudar o loop

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+        }
+    }
+
+    (noMissingDependencies) ? m_status->allFine() : m_status->somethingIsMissing();
```
Não colocar operador ternário perdido no meio do nada

CarlosCraveiro replied:
Alterado para um if/else statement!!

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+    case Status::STOPPED:
+    case Status::UNKNOWN:
+    case Status::FINISHED:
+        std::cout << "ERROR!! This was not supose to happen!! - Static Service" << std::endl;
```
Lançar exceção

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+        }
+    }
+
+    (noMissingDependencies) ? m_status->allFine() : m_status->somethingIsMissing();
```
Cópia de código da outra função `autoUpdate`

CarlosCraveiro replied:
Retirado dos proxys filhos e implementado no proxy pai, aka Service Proxy!!

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+    case Status::UNINITIALIZED:
+    case Status::UNKNOWN:
+        std::cout << "ERROR!! This was not supose to happen!! - Routine Service" << std::endl;
+    };
```
Mais uma função repetida

CarlosCraveiro replied:
Essa função foi melhorada com o uso de smart pointers, mass quanto à implementação de uma função comum no proxy pai para ser utilizada pelos proxys filhos, esta exigiria de uma reformulação do uso dos states pelas classes service... podemos considerar como um "Problema em aberto da computação".

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,243 @@
+
+    return (nlohmann::json) { { "serviceId", m_realServiceId }, { "State", currStatus } };
+}
+
```
Função repetida

CarlosCraveiro replied:
Função implementada no proxy pai, aka ServiceProxy, e retirada dos proxys filhos.

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,256 @@
+        : m_state(state)
+        , m_statusMtx(statusMtx)
+    {
+    }
```
Coloca o construtor em cima

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,256 @@
+
+class ServiceProxy {
+public:
+    enum proxyT { STATIC_SERVICE = 0,
```
Outro nome de enum pouco sugestivo

CarlosCraveiro replied:
Nome trocado para proxy_t.

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,256 @@
+protected:
+    class ProxyConfigs {
+    public:
+        struct Dependencie {
```
A forma correta de escrita é `Dependency`

CarlosCraveiro replied:
Erro gramatical corrigido!!

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,256 @@
+    std::mutex m_statusMtx;
+
+public://temporary
+    ProxyConfigs m_proxyConfigs;
```
Retirar coisas temporárias

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,256 @@
+    virtual void serviceCycle() { }
+
+protected:
+
```
Vazio

mairacanal reviewed (changes requested):
Eu não fui capaz de terminar a revisão, mas, no geral, eu recomendo você mesmo dar uma limpeza geral nos arquivos, ver o que não tá sendo utilizado e etc. Além disso, eu acho que alguns arquivos não foram formatados.

CarlosCraveiro committed:
IService: renames directory and erases empty .cpp

Due to the fact that "IService" is just a interface it doesn't need a
.ccp file, also  the directory "service/" and the file "service.hpp"
were renamed to have the same name of "IService".

CarlosCraveiro committed:
StaticService & RoutineService: creates both

Creates the classes StaticService and RoutineService in file
iservice.hpp, both were inherited from iservice, located in
the same file.

The idea is to allow function overloading on the method
"deploy" at the daemon class, when the user decides to choose between
the two options.

CarlosCraveiro committed:
ServiceProxy: splits the class and ties up loose ends

This commit will be followed by other two, because the gigantic file
service-proxy.hpp and its .cpp was cut each one in three.

Basically the classes staticService::Proxy and routineService::Proxy and
their states now exists in their own files. The staticService::Proxy now
is StaticServiceProxy and routineService::Proxy RoutineServiceProxy,
the external namespaces was erased and now the states uses the namespace
of their own classes.

Besides these changes, the new sliced files had their "loose ends" tied
up, with "loose ends" meaning the several "public", "protected" tags
that haunted the three classes before, despises the other several
"//Temporary!" comments which made the life of code reviewers an endless
nightmare. The solution for all these problems came from turning the
inherited proxys friends of ServiceHandler, and those friends of an
test class that I uses as an "unit test" class, so I could test the
features without spreading MacGyvers through the classes.

CarlosCraveiro committed:
RoutineServiceProxy: creates the class

This commits adds the files routine-service-proxy.hpp and its .cpp
with the declaration and definition of the class RoutineServiceProxy and
its states. Just to be more explicit, RoutineServiceProxy is the old
routineService::Proxy class with some improvements.

Some of news are the adoption of modern iterators (range based "fors")
"for(auto& [key, value] : thingtobeiterated)" and the adoption of
smart pointers.

CarlosCraveiro committed:
StaticServiceProxy: creates the class

This commits adds the files static-service-proxy.hpp and its .cpp
with the declaration and definition of the class StaticServiceProxy and
its states. Just to be more explicit, StaticServiceProxy is the old
staticService::Proxy class with some improvements.

Some of news are the adoption of modern iterators (range based "fors")
"for(auto& [key, value] : thingToBeIterated)" and the adoption
of smart pointers.

CarlosCraveiro committed:
ServiceHandler: improves syntax and fits the changes

Adoption of modern iterators for std::maps and smart-pointers, besides
fitting the changes of the "Service Proxys" classes.

CarlosCraveiro committed:
ServiceProxy: adds the standard dependency "THIS"

The special dependency "THIS", that all proxys now have, allows the
daemon class to choose the best time to start/expose their services on
the DBUS. Now when the daemon are ready to expose its services, it
changes the dependency "THIS" to state "RUNNING", so if "THIS" is the
only dependency of a specific service, now it can starts running.

CarlosCraveiro committed:
ServiceHandler: adds the method run, overloads build method

The method run changes the dependency "THIS" to state
RUNNING in all proxys handled by a ServiceHandler.

Also, now the method buildProxy instead of requiring an argument with
the "proxy-type", it detects the "proxy-type" by function overload.

CarlosCraveiro committed:
ServiceProxy: Improves clarity and reduces code repetition

To improve the clarity of the code, some grammar errors were fixed, like
"dependency" instead of "dependencie", some ternary operators were
converted in if/else statements, the names of the enums' have been changed
to make clearer their meaning and last but not least, some functions
from StaticServiceProxy and RoutineServiceProxy were transferred to
ServiceProxy.

CarlosCraveiro committed:
StaticServiceProxy: improves clarity of the code

Besides fitting the changes made on the ServiceProxy, two functions
that were repeated in the two child proxys, "reportState" and "checkState"
were transfered to the parent proxy.

CarlosCraveiro committed:
RoutineServiceProxy: improves clarity of the code

Besides fitting the changes made on the ServiceProxy, two functions
that were repeated in the two child proxys, "reportState" and "checkState"
were transfered to the parent proxy.

CarlosCraveiro committed:
ServiceHandler: fits the changes on enums's name

Due to the change of the enum's name, from "stateT" to "state_t" and
from "proxyT" to "state_t", little changes like expected output type of
functions "getProxyState" and "getAllProxyState" required changes to
compile.

CarlosCraveiro committed:
StaticServiceProxy: allows only StaticServices

Now the class StaticServiceProxy allows only classes inherited from
StaticService instead of allowing classes inherited directly from
IService.

CarlosCraveiro committed:
RoutineServiceProxy: allows only RoutineServices

Now the class RoutineServiceProxy allows only classes inherited from
RoutineService instead of allowing classes inherited directly from
IService.

CarlosCraveiro committed:
ServiceHandler: brings the class' constructor to the begin

Brings the class' constructor to the begin of class' definition in order
to improve the clarity of the code.

CarlosCraveiro committed:
ServiceProxy: brings some constructors to the top

Improves clarity of the code.

CarlosCraveiro committed:
StaticServiceProxy: throws exception on the "changeState" method

Throws exception on the method "changeState" instead of outputting
an error message in the screen.

CarlosCraveiro committed:
RoutineServiceProxy: throws exception on the "changeState" method

Throws exception on the method "changeState" instead of outputting
an error message in the screen.

mairacanal commented on src/classes/iservice/iservice.hpp:
```diff
@@ -0,0 +1,78 @@
+    explicit StaticService(std::string serviceId)
+        : IService(serviceId)
+    {
+    }
```
```suggestion
    explicit StaticService(std::string serviceId)
        : IService(serviceId) {};
```

mairacanal replied:
Mudança análoga pode ser feita em todo o arquivo.

CarlosCraveiro replied:
Mudança análoga aplicada em todo os arquivos!!

mairacanal commented on src/classes/routine-service-proxy/routine-service-proxy.cpp:
```diff
@@ -0,0 +1,76 @@
+
+void RoutineServiceProxy::MissingDependencies::allFine()
+{
+    // Expose/Hide the endpoint on the DBUS and THEN..
```
Não coloque comentário aqui. Faça o Javadoc apropriado da função.

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/routine-service-proxy/routine-service-proxy.cpp:
```diff
@@ -0,0 +1,76 @@
+
+void RoutineServiceProxy::Running::somethingIsMissing()
+{
+    // Unexpose/Hide the endpoint on the DBUS and THEN..
```
Não coloque comentário aqui. Faça o Javadoc apropriado da função.

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/routine-service-proxy/routine-service-proxy.cpp:
```diff
@@ -0,0 +1,76 @@
+
+void RoutineServiceProxy::weave()
+{
+    const std::lock_guard<std::mutex> lock(m_updateMtx);
```
```suggestion
    const std::lock_guard<std::mutex> lock{m_updateMtx};
```

Em geral, isso melhora a legibilidade do código.

CarlosCraveiro replied:
Legibilidade melhorada!!

mairacanal commented on src/classes/routine-service-proxy/routine-service-proxy.hpp:
```diff
@@ -0,0 +1,78 @@
+            : m_upperProxy(upperProxy)
+            , ServiceState(state)
+        {
+        }
```
```suggestion
        RoutineState(state_t state, RoutineServiceProxy& upperProxy)
            : m_upperProxy(upperProxy)
            , ServiceState(state) {};
```

Aplicar de forma análoga em todo arquivo

CarlosCraveiro replied:
Aplicado!!

Grillo-0 commented on src/classes/iservice/iservice.hpp:
```diff
@@ -0,0 +1,78 @@
+
+#pragma once
+#include <iostream>
```
Esse iostream desnecessário

CarlosCraveiro replied:
Resolvido!!

CarlosCraveiro committed:
CMakeLists: Allows compilation

Hot fix to allows compilation of service-classes.

Grillo-0 commented on src/classes/iservice/iservice.hpp:
```diff
@@ -0,0 +1,78 @@
+    std::string m_serviceId;
+
+protected:
+    //	DBusHandler::DBusHandler m_dbus;
```
Linha comentada

CarlosCraveiro replied:
Resolvido!!

Grillo-0 commented on CMakeLists.txt:
```diff
@@ -47,23 +47,24 @@ endmacro(ZFKD_CREATE_DICT)
 	service-handler
 	service-proxy
+    static-service-proxy
+    routine-service-proxy
```
Apesar de esse cmake morrer daqui a pouco, eu não acho legal deixar alguns arquivo comentados. Então só deixa os arquivos novos, também a identação ta errada

CarlosCraveiro replied:
Resolvido!!

Grillo-0 commented on CMakeLists.txt:
```diff
@@ -47,23 +47,24 @@ endmacro(ZFKD_CREATE_DICT)
-	locked-storage	
+    #dbus-handler
+    #config-handler
+    #locked-storage	
```
mesma coisa

CarlosCraveiro replied:
Resolvido!!

Grillo-0 commented on CMakeLists.txt:
```diff
@@ -47,23 +47,24 @@ endmacro(ZFKD_CREATE_DICT)
-	daemon
-	service
+    #daemon
+    #service
```
mesma coisa

Grillo-0 commented on src/classes/iservice/iservice.hpp:
```diff
@@ -0,0 +1,78 @@
+   * @param serviceId std::string that is the id related
+   * 		to the instantiated class.
+   */
+
```
Espaço desnecessário

CarlosCraveiro replied:
Resolvido!!

mairacanal commented on src/classes/iservice/iservice.hpp:
```diff
@@ -0,0 +1,78 @@
+
+#pragma once
+#include <iostream>
+#include <string>
```
O uso dessas bibliotecas é realmente necessário?

CarlosCraveiro replied:
Pelo menos uma, mantive só a string por motivos de clareza.

mairacanal commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,52 @@
+    return allStates;
+}
+
+void ServiceHandler::run() {
+    
+    for (auto& [serviceId , proxy] : m_serviceMap) {
+        proxy->m_proxyConfigs.changeDep("THIS", ServiceProxy::ServiceState::RUNNING);
+        proxy->autoUpdate();
+    }
+}
```
Não rodou o clang-format nesse trecho

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,57 @@
+#include "service-proxy.hpp"
+#include <iostream>
```
Não adicionar bibliotecas no cpp

CarlosCraveiro replied:
Removidas!!

Grillo-0 commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void ServiceHandler::buildServiceProxy(StaticService& userService)
+{
+        m_serviceMap.emplace(std::make_pair(
+                std::string(userService.m_serviceId),
+                std::make_unique<StaticServiceProxy>(userService, m_proxyDepsMap.at(userService.m_serviceId))));
```
Identação

CarlosCraveiro replied:
Resolvida!!

Grillo-0 commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,52 @@
+
+void ServiceHandler::buildServiceProxy(RoutineService& userService)
+{
+        m_serviceMap.emplace(std::make_pair(
+                std::string(userService.m_serviceId),
+                std::make_unique<RoutineServiceProxy>(userService, m_proxyDepsMap.at(userService.m_serviceId))));
```
Identação

CarlosCraveiro replied:
Resolvido!!

Grillo-0 commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,52 @@
+}
+
+void ServiceHandler::run() {
+    
```
Espaço desnecessário

CarlosCraveiro replied:
Resolvido!!

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.cpp:
```diff
@@ -0,0 +1,60 @@
+
+void StaticServiceProxy::Uninitialized::somethingIsMissing()
+{
+    // Unexpose/Hide the endpoint on the DBUS and THEN..
```
Não coloque comentário aqui. Faça o Javadoc apropriado da função.

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.cpp:
```diff
@@ -0,0 +1,60 @@
+
+void StaticServiceProxy::StandBy::somethingIsMissing()
+{
+    // Unexpose/Hide the endpoint on the DBUS and THEN...
```
Não coloque comentário aqui. Faça o Javadoc apropriado da função.

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.cpp:
```diff
@@ -0,0 +1,60 @@
+#include "static-service-proxy.hpp"
+
+StaticServiceProxy::StaticServiceProxy(StaticService& realService, std::map<std::string, ServiceState::state_t> depsMap)
+    : m_runnedOnce(false)
+    , ServiceProxy(realService, STATIC_SERVICE, depsMap)
```
```suggestion
StaticServiceProxy::StaticServiceProxy(StaticService& realService, std::map<std::string, ServiceState::state_t> depsMap)
    : m_runnedOnce{false}
    , ServiceProxy{realService, STATIC_SERVICE, depsMap}
```

Grillo-0 replied:
Isso é só estético? Sempre tive essa duvida

CarlosCraveiro replied:
Resolvido!!

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.cpp:
```diff
@@ -0,0 +1,60 @@
+
+void StaticServiceProxy::StaticServiceProxy::changeState(ServiceState::state_t newState)
+{
+    const std::lock_guard<std::mutex> lock(m_stateMtx);
```
```suggestion
    const std::lock_guard<std::mutex> lock{m_stateMtx};
```

CarlosCraveiro replied:
Feito!!

Grillo-0 commented on src/classes/service-handler/service-handler.hpp:
```diff
@@ -0,0 +1,28 @@
+public:
+    explicit ServiceHandler(nlohmann::json servicesConfigs);
+
+protected:
+    friend class Tester;
+
```
Esse Tester n é do seu "unit test" que nem vai ser subido?

CarlosCraveiro replied:
Sim, apagado!

Grillo-0 commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,57 @@
+#include "service-proxy.hpp"
+#include <iostream>
```
Include denecessário

CarlosCraveiro replied:
De fato!!

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.hpp:
```diff
@@ -0,0 +1,66 @@
+        StaticServiceProxy& m_upperProxy;
+
+    public:
+        StaticState(state_t state, StaticServiceProxy& upperProxy)
+            : ServiceState(state)
+            , m_upperProxy(upperProxy)
+        {
+        }
```
```suggestion
        StaticState(state_t state, StaticServiceProxy& upperProxy)
            : ServiceState{state}
            , m_upperProxy{upperProxy} {};
```

CarlosCraveiro replied:
Resolvido!!

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.hpp:
```diff
@@ -0,0 +1,66 @@
+        void somethingIsMissing() override;
+    };
+
+    bool m_runnedOnce;
```
Não deixe atributos públicos

Math-42 replied:
Na verdade ele está como protected, porem na ordem acabou ficando confuso, eu sugiro que coloque ele antes da declaração das classes.

CarlosCraveiro replied:
Feito!!

Grillo-0 commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,86 @@
+class ServiceProxy { 
+protected:
+    friend class ServiceHandler;
+    friend class Tester;
```
Já falado

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,57 @@
+
+    if (noMissingDependencies) {
+        m_state->allFine();
+
```
Apagar linha

CarlosCraveiro replied:
Resolvido!!

mairacanal commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,57 @@
+
+auto ServiceProxy::ServiceProxy::checkState() -> ServiceState::state_t
+{
+    const std::lock_guard<std::mutex> lock(m_stateMtx);
```
```suggestion
    const std::lock_guard<std::mutex> lock{m_stateMtx};
```

CarlosCraveiro replied:
Resolvido!!

mairacanal commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,86 @@
+
+        const state_t m_state;
+
+        explicit ServiceState(state_t state)
+            : m_state(state)
+        {
+        }
+
```
```suggestion
        explicit ServiceState(state_t state)
            : m_state{state} {};

```

CarlosCraveiro replied:
Resolvido!!

Grillo-0 commented on src/classes/static-service-proxy/static-service-proxy.hpp:
```diff
@@ -0,0 +1,66 @@
+
+protected:
+    friend class ServiceHandler;
+    friend class Tester;
```
Já falado

mairacanal commented on src/classes/service-handler/service-handler.hpp:
```diff
@@ -0,0 +1,28 @@
+#pragma once
+#include "../iservice/iservice.hpp"
```
Essa biblioteca é realmente necessária?

CarlosCraveiro replied:
Por motivos de clareza sim, pois nela são definido os services

mairacanal reviewed (changes requested):
Passar clang-format em tudo

Math-42 commented on src/classes/iservice/iservice.hpp:
```diff
@@ -0,0 +1,78 @@
+#include <string>
+class IService {
+public:
+    /* Id of the service */
```
comentário desnecessário visto que o nome da variável já é auto explicativo

CarlosCraveiro replied:
Resolvido!!

Math-42 commented on src/classes/routine-service-proxy/routine-service-proxy.cpp:
```diff
@@ -0,0 +1,76 @@
+        m_state = std::make_unique<Finished>(*this);
+        break;
+    default:
+        throw std::logic_error("Routine Services should not have other States!");
```
Isso é muito improvável de acontecer, sendo assim poderia lançar algo como unknow status, e imprimir o valor que foi lido, apenas para elucidar melhor o log de erros.

CarlosCraveiro replied:
Feito!!

Math-42 commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,57 @@
+
+auto ServiceProxy::ServiceProxy::reportState() -> nlohmann::json
+{
+    ServiceState::state_t currStatus = checkState();
+
+    return (nlohmann::json) { { "serviceId", m_realServiceId }, { "State", currStatus } };
```
isso tudo não poderia ser resumido em uma linha só? 
```suggestion
    return (nlohmann::json) { { "serviceId", m_realServiceId }, { "State", checkState(); } };
```

CarlosCraveiro replied:
Resolvido!!

Math-42 commented on src/classes/static-service-proxy/static-service-proxy.cpp:
```diff
@@ -0,0 +1,60 @@
+        m_state = std::make_unique<StandBy>(*this);
+        break;
+    default:
+        throw std::logic_error("Static Services should not have other States!");
```
Ver a observação que fiz anteriormente no arquivo semelhante.

CarlosCraveiro replied:
Resolvido!!

CarlosCraveiro committed:
IService: improves clarity of the code

Empty spaces were removed and syntax were changed to improve code
aesthetic and improve the clarity of the code.

CarlosCraveiro committed:
ServiceProxy: improves clarity of the code

Empty spaces were removed and syntax were changed to improve code
aesthetic and the clarity of the code .

CarlosCraveiro committed:
RoutineServiceProxy: improves clarity of the code

Empty spaces were removed, syntax and error messages were changed
to improve code aesthetic and the clarity of the code.

CarlosCraveiro committed:
StaticServiceProxy: improves clarity of the code

Empty spaces were removed, syntax and error messages were changed
to improve code aesthetic and the clarity of the code.

CarlosCraveiro committed:
ServiceHandler: improves clarity of the code

Empty spaces were removed and syntax were changed
to improve code aesthetic and the clarity of the code.

CarlosCraveiro committed:
CMakeLists: Erases commented lines

Some unused classes were commented because they're not used on the
compilation of the "service-classes", such as "daemon" and
"config-handler", so this commit erases this lines to improve clarity.

Grillo-0 commented on CMakeLists.txt:
```diff
@@ -47,23 +47,18 @@ endmacro(ZFKD_CREATE_DICT)
 
 set(ZFKD_UTILITIES
 	
-	dbus-handler
-	config-handler
-	locked-storage	
```
Não deletar

CarlosCraveiro replied:
Então...mass se eu não apagar, preciso deixar comentada as linhas

CarlosCraveiro replied:
Caso contrário, o cmake não gera os arquivos, pq ele não acha os arquivos do dbus-handler e locked-storage

CarlosCraveiro replied:
config-handler ele achou, mass acho inclusive que essa classe não existe mais/ foi substituída por outra.

CarlosCraveiro replied:
Resolvido!!

Grillo-0 commented on CMakeLists.txt:
```diff
@@ -47,23 +47,18 @@ endmacro(ZFKD_CREATE_DICT)
 
 set(ZFKD_PUBLIC_CLASSES
 	
-	daemon
-	service
```
Não deletar

CarlosCraveiro replied:
Service precisa apagar sim na real, pq esse arquivo acho que não faz mais sentido com o projeto.

CarlosCraveiro replied:
Deletado só service que não está mais entre nós!

Grillo-0 commented on CMakeLists.txt:
```diff
@@ -47,23 +47,18 @@ endmacro(ZFKD_CREATE_DICT)
-	service
 	service-handler
 	service-proxy
+    static-service-proxy
+    routine-service-proxy
```
indentação

CarlosCraveiro replied:
Não tinha visto a nota de identação na requisição de mudanças passada, dsclp.

CarlosCraveiro replied:
Resolvido

Grillo-0 commented on src/classes/service-proxy/service-proxy.cpp:
```diff
@@ -0,0 +1,53 @@
+
+void ServiceProxy::ProxyConfigs::changeDep(std::string dependencieId, ServiceState::state_t currState)
+{
+    m_depsMap[dependencieId].m_currState = currState;
```
Acho que seria melhor usar o .at() em vez da sobrecarga de operador

CarlosCraveiro replied:
Qual o motivo exatamente?

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/service-handler/service-handler.cpp:
```diff
@@ -0,0 +1,53 @@
+
+void ServiceHandler::run()
+{
+
```
Remover essa linha

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/routine-service-proxy/routine-service-proxy.hpp:
```diff
@@ -0,0 +1,65 @@
+            : RoutineState { FINISHED, upperProxy } {};
+    };
+
+    std::thread m_thread;
+    std::mutex m_updateMtx;
```
Coloca as variáveis antes das classes, com o objetivo de melhorar a legibilidade

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.hpp:
```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include "../service-proxy/service-proxy.hpp"
+#include <memory>
```
Essa biblioteca é necessária? Eu não vi nenhum `unique_ptr` nessa classe.

CarlosCraveiro replied:
Então, o uso de unique pointers está no .cpp dela, mais precisamente nos "std::make_unique" do método changeState, deixei no header a inclusão mesmo só sendo usada no .cpp, pois foi pedido para não incluir as bibliotecas no .cpp.

mairacanal replied:
Deixei passar o `make_unique`. Perdão, desconsidere essa sugestão.

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.hpp:
```diff
@@ -0,0 +1,55 @@
+
+class StaticServiceProxy : public ServiceProxy {
+public:
+    StaticServiceProxy(StaticService& realService, std::map<std::string, ServiceState::state_t> depsMap);
```
Faltou importar a biblioteca `map`

CarlosCraveiro replied:
Feito!!

mairacanal commented on src/classes/static-service-proxy/static-service-proxy.cpp:
```diff
@@ -0,0 +1,74 @@
+
+void StaticServiceProxy::StaticServiceProxy::changeState(ServiceState::state_t newState)
+{
+    const std::lock_guard<std::mutex> lock { m_stateMtx };
```
Faltou importar a biblioteca `mutex`

CarlosCraveiro replied:
Feito!!

mairacanal reviewed (changes requested):
Em geral, tá excelente. Só tenho esses comentários pequenos dessa vez. Uma última pergunta: tu passou o valgrind nisso?

Grillo-0 commented on src/classes/service-proxy/service-proxy.hpp:
```diff
@@ -0,0 +1,79 @@
+#pragma once
+#include "../iservice/iservice.hpp"
+#include <iostream>
```
iostream não me parece necessário

CarlosCraveiro replied:
Verdade, parei para pensar e não tenho como justificar ele.

CarlosCraveiro replied:
Removido!!

Grillo-0 reviewed (changes requested):
Código muito bom, só alguns comentários de legibilidade e do cmake

mairacanal reviewed (dismissed):
Já deu de reclamar do código né

Toma meu Approve aí e seja feliz.

Grillo-0 reviewed (approved):
Finalmente tudo ta certo 🥳
